client/devdocs/design/search-collection.jsx

```diff
@@ -31,49 +34,106 @@ const shouldShowInstance = ( example, filter, component ) => {
+		component: PropTypes.string,
+		examplesToMount: PropTypes.number,
+		filter: PropTypes.string,
+		section: PropTypes.string,
+	};
+
+	static defaultProps = {
+		examplesToMount: 7,
+		section: 'design',
+	};
+
+	_setLazyRenderRef = ref => {
+		this._lazyRender = ref;
+	};
+
+	componentDidMount = () => {
+		this._lazyRenderTimeout = setTimeout( () => {
+			// HACK: Reach into the <LazyRender> component and force all
+			// examples to be mounted after a few seconds of the component
+			// being mounted. This means if the user loads the page and
+			// changes context the rest of the content won't wait to load
+			// until they scroll down.
+			this._lazyRender.stopListening();
+			this._lazyRender.setState( { hasBeenScrolledIntoView: true } );
+		}, 2000 );
+	};
+
+	componentWillUnmout = () => {
+		clearTimeout( this._lazyRenderTimeout );
+	};
+
+	render() {
+		const { children, component, examplesToMount, filter, section } = this.props;
+
+		let showCounter = 0;
+		const summary = [];
+
+		const examples = React.Children.map( children, example => {
+			if ( ! example || ! shouldShowInstance( example, filter, component ) ) {
+				return null;
+			}
+
+			const exampleName = getComponentName( example );
+			const exampleLink = `/devdocs/${ section }/${ camelCaseToSlug( exampleName ) }`;
+
+			showCounter++;
+
+			if ( filter ) {
+				summary.push(
+					<span key={ `instance-link-${ showCounter }` } className="design__instance-link">
+						<a href={ exampleLink }>{ exampleName }</a>
+						,&nbsp;
+					</span>
+				);
+			}
+
+			return (
+				<DocsExampleWrapper name={ exampleName } unique={ !! component } url={ exampleLink }>
+					{ example }
+					{ component && <ReadmeViewer readmeFilePath={ example.props.readmeFilePath } /> }
+				</DocsExampleWrapper>
 			);
-		}
+		} );
 
 		return (
-			<DocsExampleWrapper name={ exampleName } unique={ !! component } url={ exampleLink }>
-				{ example }
-				{ component && <ReadmeViewer readmeFilePath={ example.props.readmeFilePath } /> }
-			</DocsExampleWrapper>
-		);
-	} );
-
-	return (
-		<div className="design__collection">
-			{ showCounter > 1 &&
-				filter && (
-					<div className="design__instance-links">
-						<span>Showing </span>
-						{ summary }...
-					</div>
+			<div className="design__collection">
+				{ showCounter > 1 &&
+					filter && (
+						<div className="design__instance-links">
+							<span>Showing </span>
+							{ summary }...
+						</div>
+					) }
+				{ /*
+					The entire list of examples for `/devdocs/blocks` and
+					`/devdocs/design` takes a long time to mount, even when
+					loaded, so we render just the first few components so
+					the page change feels faster. The rest of the components
+					are loaded either once scrolled to or in a few seconds
+					after `componentDidMount()` is called.
+				*/ }
+				{ examples.length <= examplesToMount ? (
+					examples
+				) : (
+					<React.Fragment>
+						{ examples.slice( 0, examplesToMount ) }
```

Comment:
This could well be a useful component type to have abstracted if we have other large pages that could be progressively mounted. There's also a few creative solutions we could take to split the sections into smaller chunks than just "the first seven" and the "all the rest" but I started with this to see what people thought of the approach. It ends up *feeling* a lot faster even if technically the component is a bit more complicated under the hood.

Reply:
If this approach is cool with people it might be worth adding that actually, as the smaller DOM changes should feel more performant and not thrash the browser as much. I can probably add it without too much fuss.

Reply:
This starts to sounds more and more like infinite scroll, or am I missing something?

Reply:
It is in a sense, but it's specific to rendering existing content (large amounts of it), rather than waiting on content to load based on an asynchronous action like an API request.

Reply:
It could be that we don't want to bother with the scrolling and simply load more content after the timeout so the render is gradual (which thanks to React's virtual DOM diffing actually thrashes less than loading everything at once).

It's sort of hacky but it definitely thrashes less and *feels* more immediate.